p2p/state.py

```diff
@@ -81,23 +82,29 @@ def msg_queue_maxsize(self) -> int:
         # now.
         return 2000
 
-    async def _get_idle_peers(self) -> List[ETHPeer]:
-        # FIXME: Should probably use get_peers() and pass the TD of our head? It's not really
-        # necessary because peers that are behind us may very well have the trie nodes we want.
-        peers = set([cast(ETHPeer, peer) async for peer in self.peer_pool])
-        return list(peers.difference(self._peers_with_pending_requests))
+    def deregister_peer(self, peer: BasePeer) -> None:
+        # Use .pop() with a default value as it's possible we never requested anything to this
+        # peer or it had all the trie nodes we requested, so there'd be no entry in
+        # self._peer_missing_nodes for it.
+        self._peer_missing_nodes.pop(cast(ETHPeer, peer), None)
 
-    async def get_idle_peer(self) -> ETHPeer:
-        idle_peers = await self._get_idle_peers()
-        while not idle_peers:
-            await self.wait(asyncio.sleep(0.02))
-            idle_peers = await self._get_idle_peers()
-        return secrets.choice(idle_peers)
+    async def get_peer_for_request(self, node_keys: Set[Hash32]) -> ETHPeer:
+        """Return an idle peer that may have any of the trie nodes in node_keys."""
+        async for peer in self.peer_pool:
+            peer = cast(ETHPeer, peer)
```

Comment:
Do we actually need that or can we instead have a separate `PeerPool` instance for every peer type, with separate services that subscribe only to the pool they need?

Reply:
:+1: 

This seems like a good idea, since we're likely to also have different policies on how we treat different peers and we won't want to conflate those things.